Update Analytics setup to display all available accounts and properties.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #9348

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.
- [ ] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

aaemnnosttv commented on assets/js/modules/analytics-4/utils/account.js:
```diff
@@ -67,3 +67,48 @@ export function getAccountDefaults(
+ * @param {string} idKey   Key of the account ID.
+ * @return {Object} Account object with an appended ID.
+ */
+export const appendAccountID = ( account, idKey = 'account' ) => {
```
1. Is the `idKey` necessary? From what I can see, it only has a single instance in use where it uses the same key. Let's remove it if not needed. Even in PHP I only see a single use of `filter_account_with_ids` so it's probably a leftover from before at some point.
2. Regarding the name, thank you for proposing an alternate to the previous "filter", however I think we can improve it a bit more. "append" implies order which is not really applicable for an object.  
How about **`populate`** instead? i.e. `populateAccountID`. `populateAccountWithIDs` would be similar to the existing naming and be fine with me as well.

aaemnnosttv commented on assets/js/modules/analytics-4/utils/account.js:
```diff
@@ -67,3 +67,48 @@ export function getAccountDefaults(
+ * @param {string} idKey    Key of the property ID.
+ * @return {Object} Property object with appended IDs.
+ */
+export const appendPropertyAndAccountIds = ( property, idKey = 'property' ) => {
```
Same as above. I think `populatePropertyWithIDs` similar to what we have now would be a bit more clear. When I first read this I imagined you'd have a `appendPropertyID` function and this was calling them both but that is not the case.

aaemnnosttv commented on assets/js/modules/analytics-4/utils/account.js:
```diff
@@ -67,3 +67,48 @@ export function getAccountDefaults(
+ * @return {Object} Property object with appended IDs.
+ */
+export const appendPropertyAndAccountIds = ( property, idKey = 'property' ) => {
+	const obj = { ...property };
```
Rather than cloning and then mutating, I'd suggest using variables for the properties or just a partial object and then returning a new object at the end with spread, 
e.g. ` return { ...property, _id, _accountID }`

This also avoids the need for an ambiguous `obj` variable (which is good to avoid) 👍

aaemnnosttv commented on assets/js/modules/analytics-4/utils/account.js:
```diff
@@ -67,3 +67,48 @@ export function getAccountDefaults(
+export const appendPropertyAndAccountIds = ( property, idKey = 'property' ) => {
+	const obj = { ...property };
+
+	const propertyMatches = property[ idKey ]?.match( /properties\/([^/]+)/ );
+	if ( propertyMatches ) {
+		obj._id = propertyMatches[ 1 ];
+	}
+
+	const accountMatches = property.parent?.match( /accounts\/([^/]+)/ );
+	if ( accountMatches ) {
+		obj._accountID = accountMatches[ 1 ];
+	}
```
I don't think there's anything wrong with including the properties in the result unconditionally, but I realize you're following the existing implementation. It should be fine to have properties present in the output with `null` or `undefined`, we can make that change to keep things simpler here as well and remove unnecessary conditions.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -268,6 +293,28 @@ const baseReducer = ( state, { type } ) => {
+			const accountSummaries = [ ...state.accountSummaries ].map(
+				( account ) => {
+					const obj = appendAccountID( account, 'account' );
+					obj.propertySummaries = account.propertySummaries.map(
```
This broke for me when testing on an account with no properties (calling `map` on `undefined`).

aaemnnosttv commented on assets/js/util/case-insensitive-sort.js:
```diff
@@ -0,0 +1,35 @@
+ * @return {Array} The sorted list.
+ */
+export function caseInsensitiveListSort( listToSort, orderBy ) {
+	return listToSort.sort( ( a, b ) => {
+		const nameA = a[ orderBy ]?.toLowerCase() || '';
+		const nameB = b[ orderBy ]?.toLowerCase() || '';
+
+		return nameA.localeCompare( nameB );
+	} );
```
This utility will mutate the input list which isn't expected by a utility IMO.

aaemnnosttv commented on assets/js/util/case-insensitive-sort.js:
```diff
@@ -0,0 +1,35 @@
+ * @return {Array} The sorted list.
+ */
+export function caseInsensitiveListSort( listToSort, orderBy ) {
+	return listToSort.sort( ( a, b ) => {
```
```suggestion
	return [ ...listToSort ].sort( ( a, b ) => {
```

aaemnnosttv commented on includes/Modules/Analytics_4.php:
```diff
@@ -1048,7 +1048,12 @@ protected function create_data_request( Data_Request $data ) {
+				return $this->get_service( 'analyticsadmin' )->accountSummaries->listAccountSummaries(
+					array(
+						'pageSize'  => 200,
+						'pageToken' => $data['pageToken'] ?? '',
```
Does this break if `null` is passed as a `pageToken`? If not, then the `??` isn't necessary here since `$data` is a `Data_Request` which implements array access, and returns any undefined keys as `null` 👍 See `Data_RequestTest::test_array_access`

aaemnnosttv commented on includes/Modules/Analytics_4.php:
```diff
@@ -1666,23 +1671,9 @@ protected function parse_data_response( Data_Request $data, $response ) {
-				return Sort::case_insensitive_list_sort(
-					$account_summaries,
-					'displayName'
+				return array(
+					'accountSummaries' => $response->getAccountSummaries(),
+					'nextPageToken'    => $response->getNextPageToken(),
```
Since the response is essentially unchanged here, let's simply remove the case and return it as-is.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.test.js:
```diff
@@ -165,7 +166,8 @@ describe( 'modules/analytics-4 accounts', () => {
 		} );
 
 		describe( 'selectAccount', () => {
```
Please also add tests for the new `transformAndSortAccountSummaries` action.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -43,24 +43,41 @@ import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store
 const fetchGetAccountSummariesStore = createFetchStore( {
 	baseName: 'getAccountSummaries',
-	controlCallback() {
+	controlCallback( params ) {
+		const { pageToken = '' } = params || {};
```
This is often how we do it in other instances I believe.
```suggestion
	controlCallback( { pageToken } ) {
```

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -43,24 +43,41 @@ import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store
 	},
-	reducerCallback( state, accountSummaries ) {
-		return { ...state, accountSummaries };
+	argsToParams: ( args ) => {
+		return { pageToken: args?.pageToken ?? '' };
```
`argsToParams` is a way to transform positional args into an object of parameters since the signature of many actions uses one or more positional args.

I think we could do the same to simplify things for the account summaries fetch action since the only parameter it has that we'll use is the pageToken, so we could update it like so
```suggestion
	argsToParams: ( pageToken ) => {
		return { pageToken };
```
This would require a change where it's called of course but could help simplify things a bit 👍 I'm not against keeping it as an object though because sometimes positional args can be hard to understand in context.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -234,6 +253,12 @@ const baseActions = {
 
 		return matchedAccount || null;
 	},
+
+	transformAndSortAccountSummaries() {
```
Please add a doc block to explain what this does.

aaemnnosttv reviewed (changes requested):
Thanks @ankitrox ! This is pretty solid and looks nice overall. I ran into a few problems in my testing which should be relatively quick to address and suggested a few other changes and feedback for your consideration. LMK if you have any questions 👍

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -234,55 +254,103 @@ const baseActions = {
+			do {
+				const { error, response } =
+					yield fetchGetAccountSummariesStore.actions.fetchGetAccountSummaries(
+						{
+							pageToken: nextPageToken,
+						}
```
This doesn't work as expected since the fetch action expects a positional argument of just the token rather than this params-like object. It works for the initial request, but won't for subsequent pages.

See 
<img width="669" alt="image" src="https://github.com/user-attachments/assets/72848f7a-8b8c-4677-9195-31dc9cd46e60">

It should be as simple as this to fix (may need formatting)
```suggestion					
						nextPageToken
```

ankitrox replied:
@aaemnnosttv If we pass `nextPageToken` instead of `{pageToken: nextPageToken}`, the create-fetch-store throws the error 

```
TypeError: Cannot convert undefined or null to object
```

`fetchGetAccountSummaries` will always be called with `{pageToken: nextPageToken}` and we won't be passing string to it.

aaemnnosttv commented on includes/Modules/Analytics_4.php:
```diff
@@ -1048,7 +1048,12 @@ protected function create_data_request( Data_Request $data ) {
-				return $this->get_service( 'analyticsadmin' )->accountSummaries->listAccountSummaries( array( 'pageSize' => 200 ) );
+				return $this->get_service( 'analyticsadmin' )->accountSummaries->listAccountSummaries(
+					array(
+						'pageSize'  => 200,
```
You can test this PR by _temporarily_ setting the `pageSize` to something small like `5` to verify multiple pages are requested (assuming you have a Google account with more than 5 GA accounts :)

aaemnnosttv reviewed (commented):
Thanks @ankitrox – this is nearly there, just a few things left to address.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/accounts.js:
```diff
@@ -43,24 +44,41 @@ import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store
 			'analytics-4',
 			'account-summaries',
-			{},
+			pageToken,
```
This also needs to be an object – it's currently just passing through the string, which is why this is being split into an object of letters in my last comment.

```suggestion
			{ pageToken },
```

ankitrox commented:
@aaemnnosttv I've addressed all the feedback points except for the last one which suggests to change `pageToken` to `{ pageToken }` and directly passing `nextPageToken` to `fetchGetAccountSummaries`. 

This is creating error in fetch store because fetch store is trying to serialise the object with `null`.

aaemnnosttv commented:
> I've addressed all the feedback points except for the last one which suggests to change `pageToken` to `{ pageToken }` and directly passing `nextPageToken` to `fetchGetAccountSummaries`.
> 
> This is creating error in fetch store because fetch store is trying to serialise the object with `null`.

@ankitrox this requires the additional comment I left in https://github.com/google/site-kit-wp/pull/9440#discussion_r1790608448 which seems to have been missed :)

I'll see if I can address it quickly as the rest seems good to go here.

aaemnnosttv commented:
Looks good now, there was a weird problem with passing `null` as an arg to the fetch store fetch action that was causing a cryptic error that wasn't obvious as to where it was coming from but that's fixed now by using an empty string instead along with the other change mentioned above. I also addressed one place where accountSummaries were not being awaited, which probably would have been fine before, but would no longer work correctly I think since accounts will be present from the first response but the matching may not find the right account if the matching account was in a subsequent request to be made.

Tested with a `pageSize` of `5` and LGTM!
<img width="724" alt="image" src="https://github.com/user-attachments/assets/d1d7e9b6-8ea7-487b-abb8-b7c61ae9d3e3">

aaemnnosttv reviewed (approved):
Great, thanks @ankitrox !
